API doc updates

alexwishart opened the pull request:
Part of work for this [issue](https://github.ibm.com/arf/planning-swift-solutions/issues/493). @helenmasters can't add you as a reviewer, so tagging you instead 🙂

Commit:
API changes

Commit:
Update README and add jazzy docs

Commit:
Update README and add jazzy docs

helenmasters commented:
Alex, as I can't add myself as a reviewer I'll put some comments here.

- I think it'd be useful to keep the basic sample usage in the README.md to minimise the number of clicks that I would need to perform as a user before finding the information I was looking for.

- In the Statistics struct I would like to see more clarity on the descriptions for hits, misses and numberOfKeys, something more like -

hits - The total number of times an access attempt successfully retrieved an item from the cache.
misses - The total number of times an access attempt did not find an item in the cache.
numberOfKeys - The number of entries currently in the cache. (I realise that this is called numberOfKeys but that is not what a user would be thinking about when using this I don't think...).

- Personally I don't like the description "The statistics of the cache." but something like "Statistics about the cache, including the number of hits and misses and the number of items in the cache." is maybe too wordy? I'm not sure please can you take another look at this.

helenmasters commented:
In the top level description the KituraCache class description I think the 'object with a Hashable key' needs some work (and the Hashable needs to be in quotes) as it's not clear if the key needs to be part of the object or not. The second sentence "Time To Live can be specified per entry, or will take a default value specified when the cache is initialised." I found confusing, maybe something like the following would be clearer

"The Time to live (TTL), i.e. the lifespan of the data, can be specified either when the cache is initialised or when an object is added."

helenmasters commented:
On the init method the description for the defaultTTL parameter doesn't make sense to me.

"The default Time to Live value (in seconds). Used for cache entries for which the TTL is not specified otherwise. If it is not specified here, the default value is 0 which maps to infinity."

I think that the second two sentences need work - we certainly don't need to talk about infinity! Something like "...the default TTL value is 0 which means that objects in the cache will not expire." might help.

Commit:
Requested fixes
Have re-added info back into README, including new badges, and made various fixes to API reference in code

alexwishart commented:
@helenmasters Have made the requested changes, plus a few others that I managed to find this afternoon. Have also added new badges and logo to the README - what do you think? 🙂

rolivieri commented:
Just added @helenmasters as the reviewer.

Commit:
Further requested fixes
Added links to API docs from README, fixed title wording and moved comments from inside code snippets to above.

Commit:
New banner & minor fixes

Commit:
Change badge link, logo size, small fix

Commit:
Fixing title

helenmasters commented on README.md:
```diff
@@ -1,75 +1,74 @@
+let cache = KituraCache()
+...
+if let item = cache.object(forKey: 1) {
+//Object with key of 1 retrieved from cache.
```
It might be clearer to indent both the comment and ... which are within the else { }

helenmasters commented on README.md:
```diff
@@ -1,75 +1,74 @@
+}
+else {
+//No object stored in cache with key of 1.
+...
```
Same comment as above regarding indentation.

helenmasters commented on README.md:
```diff
@@ -1,75 +1,74 @@
-   - *hits* the number of cache hits
-   - *misses* the number of cache misses
-   - *numberOfKeys* the total number of keys in the cache
+_Refer to [KituraCache](https://ibm-swift.github.io/Kitura-Cache/Classes/KituraCache) and [Statistics](https://ibm-swift.github.io/Kitura-Cache/Structs/Statistics) for more information and further configuration._
```
I don't think the words "and further configuration" add anything useful to this sentence - please can you remove them.

helenmasters commented on Sources/KituraCache/Statistics.swift:
```diff
@@ -16,16 +16,37 @@
     public internal(set) var misses = 0
     
-    /// The total number of keys in the cache.
+    /** The total number of entries curently in the cache.
```
curently - spelling needs fixing to currently!

helenmasters commented on Sources/KituraCache/Statistics.swift:
```diff
@@ -16,16 +16,37 @@
     
-    /// The total number of keys in the cache.
+    /** The total number of entries curently in the cache.
+     ### Usage Example: ###
```
This is displaying as "The total number of entries curently in the cache. ### Usage Example: ###" when it's loaded so needs fixing.

helenmasters commented on Sources/KituraCache/Statistics.swift:
```diff
@@ -16,16 +16,37 @@
     public internal(set) var hits = 0
     
-    /// The number of the cache misses.
+    /** The total number of times an access attempt was unable to retrieve an entry in the cache.
```
I think "from the cache" would be better than "in the cache" here.

helenmasters commented on Sources/KituraCache/KituraCache.swift:
```diff
@@ -64,12 +69,22 @@ public class KituraCache {
+    /**
+     Adds a new entry or updates the existing entry if the key is already associated with an object in the cache. The lifespan of the entry (in seconds) in the cache can be set using the optional withTTL parameter.
+     ### Usage Example: ###
+     In this example, item is an instance of a struct object with an id field which conforms to Hashable.
```
struct and Hashable should be quoted.

helenmasters commented on Sources/KituraCache/KituraCache.swift:
```diff
@@ -89,11 +104,28 @@ public class KituraCache {
+    /**
+     Retrieve an object from the cache for a specified key.
+     ### Usage Example: ###
+     In this example, item has been stored in the cache with an integer key
```
Full stop at end.

helenmasters commented on Sources/KituraCache/KituraCache.swift:
```diff
@@ -102,23 +134,75 @@ public class KituraCache {
+    /**
+     Remove objects from the cache for multiple, specified keys.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Full stop at end.

helenmasters commented on Sources/KituraCache/KituraCache.swift:
```diff
@@ -102,23 +134,75 @@ public class KituraCache {
+    /**
+     Remove an object from the cache for a specified key.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Full stop at end.

helenmasters commented on Sources/KituraCache/KituraCache.swift:
```diff
@@ -102,23 +134,75 @@ public class KituraCache {
+    /**
+     Remove objects from the cache for multiple, specified keys provided in an array.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Full stop at end.

helenmasters commented on Sources/KituraCache/KituraCache.swift:
```diff
@@ -145,11 +237,21 @@ public class KituraCache {
+    /**
+     Set the Time to Live value (in seconds) for a cache entry.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Full stop at end.

Commit:
Requested fixes

alexwishart commented:
@helenmasters have made tehrequested fixes 🙂

helenmasters commented on Sources/KituraCache/Statistics.swift:
```diff
@@ -39,12 +41,13 @@ public struct Statistics {
      let cache = KituraCache()
      ...
-     let numKeys = cache.statistics.numberOfKeys
+     let numMisses = cache.statistics.numberOfKeys
```
This line looks like it was updated in error?

helenmasters commented on Sources/KituraCache/Statistics.swift:
```diff
@@ -29,7 +30,8 @@ public struct Statistics {
     
-    /** The total number of times an access attempt was unable to retrieve an entry in the cache.
+    /**
+     The total number of times an access attempt was unable to retrieve an entry in the cache.
```
Please can you change "in the cache" to "from the cache"

helenmasters commented on Sources/KituraCache/Statistics.swift:
```diff
@@ -19,7 +19,8 @@
     
-    /** The total number of times an access attempt successfully retrieved an entry from the cache.
+    /**
+     The total number of times an access attempt successfully retrieved an entry in the cache.
```
Please can you change "in the cache" to "from the cache"

helenmasters commented on Sources/KituraCache/KituraCache.swift:
`````diff
@@ -19,26 +19,31 @@ import Dispatch
+     ````swift
+     let cache = KituraCache(defaultTTL: 3600, checkFrequency: 600)
+     ````
+     - Parameter defaultTTL: The Time to Live value (in seconds) of an entry if a TTL value is not specified when the entry is later added. If the default TTL is not defined during this initialisation, entries in the cache are configured to not expire.
`````
the entry is later added -> "an entry is added later" I think would be clearer.

Commit:
Requested docs fixes

helenmasters reviewed (approved):
The changes all look good to me.  
@rolivieri can you do a technical review please, then if you're happy we'll be good to go.

Also @rolivieri do you think us linking to an icon in the Kitura central repository for the README banner headers will cause us issues (rather than having a copy in each local repo)?

helenmasters commented:
@tunniclm I can't add you as a reviewer, so tagging you instead 🙂, please can you do a technical review so that this can be merged.

tunniclm committed:
Wording and formatting changes

tunniclm committed:
Update jazzy docs
